query/src/main/scala/filodb/query/exec/RangeVectorTransformer.scala

```diff
@@ -240,4 +283,62 @@ final case class SortFunctionMapper(function: SortFunctionId) extends RangeVecto
       source
     }
   }
-}
\ No newline at end of file
+  override def funcParams: Seq[FuncArgs] = Nil
+}
+
+final case class ScalarFunctionMapper(function: ScalarFunctionId,
+                                      timeStepParams: RangeParams) extends RangeVectorTransformer {
+  protected[exec] def args: String =
+    s"function=$function, funcParams=$funcParams"
+
+  def scalarImpl(source: Observable[RangeVector]): Observable[RangeVector] = {
+
+    val resultRv = source.toListL.map { rvs =>
+      if (rvs.size > 1) {
+        Seq(ScalarFixedDouble(timeStepParams, Double.NaN))
+      } else {
+        Seq(ScalarVaryingDouble(rvs.head.rows.map(r => (r.getLong(0), r.getDouble(1))).toMap))
+      }
+    }.map(Observable.fromIterable)
+    Observable.fromTask(resultRv).flatten
+  }
+  def apply(source: Observable[RangeVector],
+            queryConfig: QueryConfig,
+            limit: Int,
+            sourceSchema: ResultSchema,
+            paramResponse: Observable[ScalarRangeVector] = Observable.empty): Observable[RangeVector] = {
+
+    function match {
+      case Scalar => scalarImpl(source)
+      case _ => throw new UnsupportedOperationException(s"$function not supported.")
+    }
+  }
+  override def funcParams: Seq[FuncArgs] = Nil
+
+}
+
+final case class VectorFunctionMapper() extends RangeVectorTransformer {
+  protected[exec] def args: String =
+    s"funcParams=$funcParams"
+
+  def apply(source: Observable[RangeVector],
+            queryConfig: QueryConfig,
+            limit: Int,
+            sourceSchema: ResultSchema,
+            paramResponse: Observable[ScalarRangeVector] = Observable.empty): Observable[RangeVector] = {
+    val resultRv = source.toListL.map { rvs =>
```

Comment:
This can be simplified significantly.  Instead of using `toListL`, mapping over it, then converting back to an Observable, you can simply do `source.map { rv => new RangeVector{....} }`

Speaking of which, this is basically a NOP, as the new RangeVector just has the same keys and rows as the original.  So why is this transform even needed?

Reply:
Ohh yes, we don't need toListL. I will remove it.
Input to VectorFunctionMapper is a ScalarRangeVector. VectorFunctionMapper should generate a normal RangeVector